jigasi: init at 1.1-311-g3de47d0 + module

yayayayaka opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Jitsi Gateway to SIP is a server-side application that links allows
regular SIP clients to join Jitsi Meet conferences hosted by Jitsi
Videobridge.

Resolves #136911
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

yu-re-ka commented on pkgs/servers/jigasi/sip-communicator.properties:
```diff
@@ -0,0 +1,252 @@
+#Sample config with one XMPP and one SIP account configured
```
Does this file need to be committed in nixpkgs? It looks like it could just be copied from `$src` to `$out`.

yayayayaka replied:
Good catch! Will update it shortly.

SuperSandro2000 commented on pkgs/servers/jigasi/default.nix:
```diff
@@ -0,0 +1,46 @@
+{ lib, stdenv, fetchurl, dpkg, jre_headless, nixosTests }:
+
+let
+  pname = "jigasi";
+  version = "1.1-195-g65ef768";
+  arch = "amd64";
+  src = fetchurl {
+    url = "https://download.jitsi.org/stable/${pname}_${version}-1_${arch}.deb";
+    sha256 = "0fq4gb4pib5bm5bdh9qr1xszzwn8ryar0n3a32s333zq5hpv9xb4";
+  };
+in
+stdenv.mkDerivation {
+  inherit pname version src;
```
```suggestion
stdenv.mkDerivation rec {
  pname = "jigasi";
  version = "1.1-195-g65ef768";

  src = fetchurl {
    url = "https://download.jitsi.org/stable/jigasi_${version}-1_amd64.deb";
    sha256 = "0fq4gb4pib5bm5bdh9qr1xszzwn8ryar0n3a32s333zq5hpv9xb4";
  };
```

SuperSandro2000 commented on pkgs/servers/jigasi/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  dontBuild = true;
+
+  unpackCmd = "${dpkg}/bin/dpkg-deb -x $src debcontents";
```
```suggestion
  unpackCmd = "dpkg-deb -x $src debcontents";
```
Please use nativeBuildInputs instead of hard coding the path.

SuperSandro2000 commented on pkgs/servers/jigasi/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  installPhase = ''
+    runHook preInstall
+    substituteInPlace usr/share/${pname}/${pname}.sh \
```
```suggestion
    substituteInPlace usr/share/jigasi/jigasi.sh \
```

SuperSandro2000 commented on pkgs/servers/jigasi/default.nix:
```diff
@@ -0,0 +1,46 @@
+      --replace "exec java" "exec ${jre_headless}/bin/java"
+
+    mkdir -p $out/{share,bin}
+    mv usr/share/${pname} $out/share/
+    mv etc $out/
+    ln -s $out/share/${pname}/${pname}.sh $out/bin/${pname}
```
```suggestion
    mv usr/share/jigasi $out/share/
    mv etc $out/
    ln -s $out/share/jigasi/jigasi.sh $out/bin/${pname}
```

SuperSandro2000 commented on pkgs/servers/jigasi/default.nix:
```diff
@@ -0,0 +1,46 @@
+    platforms = platforms.linux;
+  };
+}
+
```
One final new line to much.

yayayayaka commented:
While fiddling with the module, I noticed this exception occur: 
```
Sep 11 15:08:58 yunix jigasi-start[1859763]: java.lang.UnsatisfiedLinkError: Directory separator should not appear in library name: /tmp/jna--116
0142225/jna500083100422037839.tmp                                                                                                                
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:827)                               
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at java.base/java.lang.System.loadLibrary(System.java:1873)                                 
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at org.jitsi.utils.JNIUtils.loadNativeInClassloader(JNIUtils.java:137)                      
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at org.jitsi.utils.JNIUtils.loadLibrary(JNIUtils.java:86)                                   
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at org.jitsi.utils.JNIUtils.loadLibrary(JNIUtils.java:45)                                   
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at net.java.sip.communicator.impl.sysactivity.SystemActivityNotifications.<clinit>(SystemAct
ivityNotifications.java:125)                                                                                                                     
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at net.java.sip.communicator.impl.sysactivity.SystemActivityNotificationsServiceImpl.isSuppo
rted(SystemActivityNotificationsServiceImpl.java:606)                                                                                            
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at net.java.sip.communicator.impl.sysactivity.SystemActivityNotificationsServiceImpl.start(S
ystemActivityNotificationsServiceImpl.java:143)                                                                                                  
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at net.java.sip.communicator.impl.sysactivity.SysActivityActivator.start(SysActivityActivato
r.java:70)                                                                                                                                       
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at org.jitsi.impl.osgi.framework.BundleImpl.start(BundleImpl.java:307)                      
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at org.jitsi.impl.osgi.framework.launch.FrameworkImpl.startLevelChanged(FrameworkImpl.java:4
72)                                                                                                                                              
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at org.jitsi.impl.osgi.framework.startlevel.FrameworkStartLevelImpl$Command.run(FrameworkSta
rtLevelImpl.java:137)                                                                                                                            
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at org.jitsi.impl.osgi.framework.AsyncExecutor.runInThread(AsyncExecutor.java:122)          
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at org.jitsi.impl.osgi.framework.AsyncExecutor.access$000(AsyncExecutor.java:28)            
Sep 11 15:08:58 yunix jigasi-start[1859763]:         at org.jitsi.impl.osgi.framework.AsyncExecutor$1.run(AsyncExecutor.java:231)
```
No idea how to fix this yet.

yayayayaka commented:
Done! Tested on my own jitsi-meet instance and got calls from jitsi-meet -> phone network and phone-network to jitsi-meet working. This is now ready to be reviewed.

yayayayaka commented on nixos/modules/services/networking/jigasi.nix:
```diff
@@ -0,0 +1,230 @@
+        File containing environment variables to be passed to the jigasi service,
+        in which secret tokens can be specified securely by defining values for
+        <literal>JIGASI_SIPUSER</literal>,
+        <literal>JIGASI_SIPPWD</literal> and
+        <literal>JIGASI_SIPSERVER</literal>.
```
While testing my setup with my own Asterisk I noticed this module is missing a way to specify the Server Port. Marking as draft again.
```suggestion
        <literal>JIGASI_SIPPWD</literal>,
        <literal>JIGASI_SIPSERVER</literal> and
        <literal>JIGASI_SIPPORT</literal>.
```

yayayayaka replied:
Resolved in f4b4a57aec1

yayayayaka commented:
Alright, it seems I have figured out how to bring the dial-in feature to work. I have tested the PR with a self-hosted Asterisk connected to my SIP Trunk. Each participant was able to each each other.

Looking forward to see this tested by someone else to confirm that my packaging is working as intended.

yayayayaka commented:
Marking as draft again in order to bring the jigasi package on par with the rest of the jitsi stack on master.

SuperSandro2000 commented on pkgs/servers/jigasi/default.nix:
```diff
@@ -0,0 +1,47 @@
+{ lib, stdenv, fetchurl, dpkg, jdk11, nixosTests }:
+
+let
+  pname = "jigasi";
+  version = "1.1-231-gd792d0f";
+  arch = "amd64";
+  src = fetchurl {
+    url = "https://download.jitsi.org/stable/${pname}_${version}-1_${arch}.deb";
+    sha256 = "0riij1i3k33rgzs4gqjffvcrx18a1dmz825nzk85xpsh972z0a4n";
+  };
+in
+stdenv.mkDerivation {
+  inherit pname version src;
```
```suggestion
stdenv.mkDerivation rec {
  pname = "jigasi";
  version = "1.1-231-gd792d0f";
  
  src = fetchurl {
    url = "https://download.jitsi.org/stable/${pname}_${version}-1_amd64.deb";
    sha256 = "0riij1i3k33rgzs4gqjffvcrx18a1dmz825nzk85xpsh972z0a4n";
  };
```
or can we easily build from source?

SuperSandro2000 commented on pkgs/servers/jigasi/default.nix:
```diff
@@ -0,0 +1,47 @@
+
+  installPhase = ''
+    runHook preInstall
+    substituteInPlace usr/share/${pname}/${pname}.sh \
+      --replace "exec java" "exec ${jdk11}/bin/java"
+
+    mkdir -p $out/{share,bin}
+    mv usr/share/${pname} $out/share/
+    mv etc $out/
+    ln -s $out/share/${pname}/${pname}.sh $out/bin/${pname}
```
Please don't substitute pname in every possible location. This just makes the code harder to read.

yayayayaka replied:
Alright, will change that.

cleeyv commented on nixos/modules/services/networking/jigasi.nix:
```diff
@@ -0,0 +1,237 @@
+        # Disable certificate verification for self-signed certificates
+        sed -i \
+          's|\(# \)\(net.java.sip.communicator.service.gui.ALWAYS_TRUST_MODE_ENABLED=true\)|\2|g' \
+          "${sipCommunicatorPropertiesFile}"
```
```suggestion
          "${sipCommunicatorPropertiesFile}"
        # Append lines from services.jigasi.config
        echo -e "${extraConfigs}" >> "${sipCommunicatorPropertiesFile}"
```

cleeyv commented on nixos/modules/services/networking/jigasi.nix:
```diff
@@ -0,0 +1,237 @@
+        "-Dnet.java.sip.communicator.SC_HOME_DIR_LOCATION" = "${stateDir}";
+        "-Dnet.java.sip.communicator.SC_HOME_DIR_NAME" = "${homeDirName}";
+        "-Djava.util.logging.config.file" = "${pkgs.jigasi}/etc/jitsi/jigasi/logging.properties";
+      };
```
I got the `services.jigasi.config` option to work by adding one line here to create a variable containing those configs and a second line at the end of the service preStart script to append those configs to the end of the jigasi sip-communicator.properties file: https://github.com/cleeyv/nixpkgs/commit/d5b7e139529ebf8bc5c66babef855b976959d0e8
```suggestion
      };
      extraConfigs = concatStringsSep "\n" (mapAttrsToList (k: v: "${k}=${v}") cfg.config);
```

cleeyv commented on pkgs/servers/jigasi/default.nix:
```diff
@@ -0,0 +1,47 @@
+{ lib, stdenv, fetchurl, dpkg, jdk11, nixosTests }:
+
+let
+  pname = "jigasi";
+  version = "1.1-231-gd792d0f";
+  arch = "amd64";
+  src = fetchurl {
+    url = "https://download.jitsi.org/stable/${pname}_${version}-1_${arch}.deb";
+    sha256 = "0riij1i3k33rgzs4gqjffvcrx18a1dmz825nzk85xpsh972z0a4n";
```
```suggestion
  pname = "jigasi";
  version = "1.1-265-g4049f97";
  arch = "all";
  src = fetchurl {
    url = "https://download.jitsi.org/stable/${pname}_${version}-1_${arch}.deb";
    sha256 = "ZFSvjSXYBUpbEc4IDm6ccdy4CF2F5dPqvzLH1H9CwzY=";
```
Update jigasi to the latest stable version.

cleeyv commented on nixos/modules/services/web-apps/jitsi-meet.nix:
```diff
@@ -265,14 +278,21 @@ in
 
       script = let
-        secrets = [ "jicofo-component-secret" "jicofo-user-secret" "jibri-auth-secret" "jibri-recorder-secret" ] ++ (optional (cfg.videobridge.passwordFile == null) "videobridge-secret");
+        secrets =
+          [ "jicofo-component-secret"
+            "jicofo-user-secret"
+            "jibri-auth-secret"
+            "jibri-recorder-secret"
+            "jigasi-user-secret"
+            "jigasi-component-secret"
+          ] ++ (optional (cfg.videobridge.passwordFile == null) "videobridge-secret");
```
When this PR branch is rebased to master there will be a merge conflict in this file. I believe it is primarily because of this recent commit implementing some of the same updates you did for the PR, but in a different way: https://github.com/NixOS/nixpkgs/commit/aea940da63c27f503915a5ca44cbb9eddf2674f6 
As part of the rebase for my transcription branch, this is how I modified your commit to merge the two versions: https://github.com/cleeyv/nixpkgs/commit/7778f69c0bb323e7cae7caf3581da7a633bd8962

SuperSandro2000 replied:
```suggestion
           ] ++ optional (cfg.videobridge.passwordFile == null) "videobridge-secret"; 
```

cleeyv reviewed (changes requested):
I successfully tested this PR by using it to deploy a Jitsi Meet instance with Jigasi configured to run transcription using the newly packaged Vosk service: https://github.com/NixOS/nixpkgs/pull/186917/
This required updating the jigasi package to the latest version, fixing the jigasi module to implement the `services.jigasi.config` option and rebasing everything to master. I've added those changes as comments in this review. Here are all the combined commits that were needed for jigasi/vosk transcription to work: https://github.com/NixOS/nixpkgs/compare/master...cleeyv:nixpkgs:transcription

yayayayaka commented on nixos/modules/services/networking/jigasi.nix:
```diff
@@ -0,0 +1,237 @@
+      '';
+    };
+
+    config = mkOption {
```
Thank you for taking a look! I will pick your changes and try to rebase against master.

yayayayaka commented:
Rebased against master and updated jigasi to the latest release. Haven't tested it yet.

yayayayaka committed:
maintainers: add yayayayaka to team jitsi

yayayayaka committed:
jigasi: init at 1.1-311-g3de47d0

Jitsi Gateway to SIP: a server-side application that allows regular SIP
clients to join Jitsi Meet conferences hosted by Jitsi Videobridge.

yayayayaka committed:
nixos/jitsi-meet: Integrate jigasi

This commit introduces the possibility to optionally enable the Jitsi
Gateway to SIP (jigasi) module. SIP credentials can be defined in
`services.jigasi.environmentFile`.

yayayayaka committed:
nixos/jigasi: init

yayayayaka commented:
Still not giving up! Tested my changes with success:

![2023-12-07T13:58:16+01:00](https://github.com/NixOS/nixpkgs/assets/73759599/040156a3-ce9e-4dc7-9f94-8a6dc47daf11)

Lassulus commented:
code looks good, nixpkgs-review seems happy, and the nixos test for jitsi-meet still works.
